node/assert.ts

```diff
@@ -317,21 +318,63 @@ function notStrictEqual(
   );
 }
 
-function deepEqual() {
+function deepEqual(
+  actual: unknown,
+  expected: unknown,
+  message?: string | Error,
+) {
   if (arguments.length < 2) {
     throw new ERR_MISSING_ARGS("actual", "expected");
   }
 
-  // TODO(kt3k): Implement deepEqual
-  throw new Error("Not implemented");
+  if (isDeepEqual(actual, expected)) {
+    return;
+  }
+
+  if (typeof message === "string") {
+    throw new AssertionError({
+      message,
+    });
+  } else if (message instanceof Error) {
+    throw message;
+  }
+
+  throw createAssertionError({
+    actual,
+    expected,
+    message,
+    operator: "deepEqual",
+    generatedMessage: true,
+  });
 }
-function notDeepEqual() {
+function notDeepEqual(
+  actual: unknown,
+  expected: unknown,
+  message?: string | Error,
+) {
   if (arguments.length < 2) {
     throw new ERR_MISSING_ARGS("actual", "expected");
   }
 
-  // TODO(kt3k): Implement notDeepEqual
-  throw new Error("Not implemented");
+  if (!isDeepEqual(actual, expected)) {
+    return;
+  }
+
+  if (typeof message === "string") {
+    throw new AssertionError({
+      message,
+    });
+  } else if (message instanceof Error) {
+    throw message;
+  }
+
+  throw createAssertionError({
+    actual,
+    expected,
+    message,
+    operator: "notDeepEqual",
+    generatedMessage: true,
+  });
```

Comment:
Is it possible to avoid all of this duplication. Look at the `innerFail()` function in Node's `assert` library as an example.